### DISCUSSION TITLE
Diverging weights after hvd.load_model

### DISCUSSION THREAD

**Discussion by gchesnokov** [horovod/horovod#2769] [Q&A]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 2.1.1
3. Horovod version: 0.18.1
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version:
8. OS and version:
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Hi! I am observing strange behaviour while loading model from checkpoint to continue training. Precisely if I don't compile tf.keras.Model loaded with hvd.load_model(), weights on different workers start to diverge. If I do compile model, training continues as expected, weights on workers are synchronised. 
[Gist](https://gist.github.com/gchesnokov/4116d234cd1524a485f9b943efd7c1cf) to reproduce (slightly modified keras mnist script from horovod examples).
Steps to reproduce:
1.  run `horovodrun -np 2 -H localhost:2 python train.py` and stop at arbitrary epoch (5 for ex.)
2. run again `horovodrun -np 2 -H localhost:2 python train.py` this will load mode from previously saved checkpoint, but will not compile it
3. run again `horovodrun -np 2 -H localhost:2 python train.py --compile-after-load` this will compile loaded model
Weights are printed to stdout.
1st step: weights are identical
2nd step: weights differ
3rd step: weights are identical again.
Basically, the question is does one need to compile model after loading or no and why.

**Comment by tgaddair** [comment]
Thanks for the repro, @gchesnokov.  Do you know if this behavior is also present in the most recent version of TF, 2.4.0?

**Comment by gchesnokov** [comment]
Tried with docker image horovod/horovod:0.20.0-tf2.3.0-torch1.6.0-mxnet1.6.0.post0-py3.7-cuda10.1. Everything seems fine, weights are synced in both setups (compile/no compile).

Unfortunately, at the moment I'm bound to lib versions mentioned in the original post. Maybe you can make an educated guess what causes such effects, so I would be able to come up with workaround (patch libs maybe) for my case.

**Comment by tgaddair** [comment]
Hey @gchesnokov, my guess is that it has to do with the `experimental_run_tf_function` param that's set when you call compile.  This flag was removed in more recent versions of TensorFlow.  I'm guessing that somewhere that is being dropped when you load the model, so perhaps you can find a way to set that flag by hand after loading?

**Comment by stale** [comment]
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.
